modyn/selector/internal/selector_strategies/downsampling_strategies/rho_loss_downsampling_strategy.py

```diff
@@ -46,24 +46,80 @@ def __init__(
+
+        model_id = self._train_il_model(next_rho_trigger_id, previous_model_id)
+        logger.info(
+            f"Trained IL model {model_id} for trigger {next_rho_trigger_id} in rho pipeline"
+            f"{self.rho_pipeline_id} with rho trigger id {next_trigger_id}."
+        )
         if self.holdout_set_strategy == "Twin":
-            raise NotImplementedError("Twin holdout set strategy is not implemented yet.")
-        self._prepare_holdout_set(next_trigger_id, selector_storage_backend)
-        self._train_il_model(next_trigger_id)
+            second_next_trigger_id = next_rho_trigger_id + 1
+            second_query = self._get_rest_data_query(self._pipeline_id, next_trigger_id)
+            self._persist_holdout_set(second_query, second_next_trigger_id, selector_storage_backend)
+            self._train_il_model(second_next_trigger_id, model_id)
+            logger.info(
+                f"Twin strategy: Trained second IL model for trigger {next_trigger_id} in rho pipeline "
+                f"{self.rho_pipeline_id} with rho trigger id {next_trigger_id}."
+            )
+        self._clean_tmp_version(self._pipeline_id, next_trigger_id, selector_storage_backend)
+
+    @staticmethod
+    def _clean_tmp_version(
+        main_pipeline_id: int, trigger_id: int, selector_storage_backend: AbstractStorageBackend
+    ) -> None:
+        assert isinstance(selector_storage_backend, DatabaseStorageBackend)
+
+        def _session_callback(session: Session) -> None:
+            session.query(SelectorStateMetadata).filter(
+                SelectorStateMetadata.pipeline_id == main_pipeline_id,
+                SelectorStateMetadata.seen_in_trigger_id == trigger_id,
+            ).update({"tmp_version": 0})
+            session.commit()
+
+        selector_storage_backend._execute_on_session(_session_callback)
+
+    @staticmethod
+    def _get_rest_data_query(main_pipeline_id: int, trigger_id: int) -> Select:
+        stmt = select(SelectorStateMetadata.sample_key).filter(
+            SelectorStateMetadata.pipeline_id == main_pipeline_id,
+            SelectorStateMetadata.seen_in_trigger_id == trigger_id,
+            SelectorStateMetadata.tmp_version == 0,
+        )
+
+        return stmt
 
     @property
     def downsampling_params(self) -> dict:
         config = super().downsampling_params
         config["rho_pipeline_id"] = self.rho_pipeline_id
-        il_model_id = self._get_latest_il_model_id(self.rho_pipeline_id, self._modyn_config)
-        assert il_model_id is not None
-        config["il_model_id"] = il_model_id
+        state = self._get_latest_rho_state(self.rho_pipeline_id, self._modyn_config)
+        assert state is not None
+        config["il_model_id"] = state[1]
         return config
 
     @staticmethod
-    def _get_latest_il_model_id(rho_pipeline_id: int, modyn_config: dict) -> Optional[int]:
+    def _get_latest_rho_state(rho_pipeline_id: int, modyn_config: dict) -> Optional[Tuple[int, int]]:
```

Comment:
Is this truly optional? doesn't look like it, seems like we assert is not None

Reply:
At the first trigger, we don't have a latest trigger id and last model id, that's when None should be returned.

Reply:
However when querying `downsampling_params`, at that point (if it is queried at the right time) it is queried from trainer server, and a TSS is already generated. So in that method we assert it to be non-None.

In method `inform_next_trigger`, the state returned could be None.

Reply:
I agree with your thought on packing! I further considered that the downside of `Tuple[Optional[int], Optional[int]]` is that, it is possible to have inconsistent None-ness, i.e. the first field being None while the second is not, and I need to double assert. That's why I chose this way.